### PR TITLE
fix(hot-module-replacement): Fix and improve HMR management API

### PR DESCRIPTION
**Pull request by sokra**
In webpack 2 `check` and `apply` return Promises.

There are more options.

### CODE REVIEW

File: src/content/api/hot-module-replacement.md [hunk 1]
```diff
@@ -93,27 +91,60 @@ module.hot.status() // Will return one of the following strings...
...
+
+The `info` object has this format:
+
+``` js
+{
+  type: "self-declined" | "declined" | 
+        "unaccepted" | "accepted" | 
+        "disposed" | "accept-errored" | 
+        "self-accept-errored" | "self-accept-error-handler-errored",
+  moduleId: 4, // The module in questions.
```
Comment (skipjack) [inline src/content/api/hot-module-replacement.md]: s/questions/question

File: src/content/api/hot-module-replacement.md [hunk 2]
```diff
@@ -93,27 +91,60 @@ module.hot.status() // Will return one of the following strings...
...
+- `ignoreUnaccepted` (boolean): Ignore changes made to unaccepted modules.
+- `ignoreDeclined` (boolean): Ignore changes made to declined modules.
+- `ignoreErrored` (boolean): Ignore errors throw in accept handlers, error handlers and while reevaulating module.
+- `onDeclined` (function(info)): Notifier for declined modules
+- `onUnaccepted` (function(info)): Notifier for unaccepted modules
+- `onAccepted` (function(info)): Notifier for accepted modules
+- `onDisposed` (function(info)): Notifier for disposed modules
+- `onErrored` (function(info)): Notifier for errors
+
+The `info` object has this format:
```
Comment (skipjack) [inline src/content/api/hot-module-replacement.md]: Might rephrase as:

> The `info` parameter will be an object containing the following values:

File: src/content/api/hot-module-replacement.md [hunk 3]
```diff
@@ -93,27 +91,60 @@ module.hot.status() // Will return one of the following strings...
...
+  moduleId: 4, // The module in questions.
+  dependencyId: 3, // For errors: the module id owning the accept handler.
+  chain: [1, 2, 3, 4], // For declined/accepted/unaccepted: the chain from where the update was propagated.
+  parentId: 5, // For declined: the module id of the declining parent
+  outdatedModules: [1, 2, 3, 4], // For accepted: the modules that are outdated and will be disposed
+  outdatedDependencies: { // For accepted: The location of accept handlers that will handle the update
+    5: [4]
+  },
+  error: new Error(...), // For errors: the thrown error
+  orginalError: new Error(...) // For self-accept-error-handler-errored:
```
Comment (skipjack) [inline src/content/api/hot-module-replacement.md]: Is this supposed to be `originalError`?

Comment (sokra) [reply src/content/api/hot-module-replacement.md]: Theoretically yes, but the typo is also in the source code. I try to fix it in the source.

**Comment by skipjack** [comment]
> Theoretically yes, but the typo is also in the source code. I try to fix it in the source.

Ah ok, want me to hold off on merging until you release a patch?

**Pull request merged**